control-plane/p4RuntimeSerializer.cpp

```diff
@@ -1412,6 +1471,90 @@ static void forAllEvaluatedBlocks(const IR::ToplevelBlock* aToplevelBlock,
     }
 }
 
+/// @return the action profile referenced in @table's implementation property,
+/// if it has one, or boost::none otherwise.
+static boost::optional<ActionProfile>
+getActionProfile(const IR::P4Table* table, ReferenceMap* refMap, TypeMap* typeMap) {
+    auto impl = table->properties->getProperty(P4V1::V1Model::instance
+                                                .tableAttributes
+                                                .tableImplementation.name);
+    if (!impl) return boost::none;
+
+    cstring name = impl->externalName();
+
+    // In P4-16, an action profile is just an instantiation of an extern. In
+    // P4-14, it's a language construct, but it's desugared into an extern
+    // instantiation in the frontend, so either way, we expect the
+    // implementation property to contain an extern constructor call.
+    if (!impl->value->is<IR::ExpressionValue>()) {
+        ::error("Table '%1%' has an implementation which is not an expression: %2%",
+                table->externalName(), impl);
+        return boost::none;
+    }
+    const auto propValue = impl->value->to<IR::ExpressionValue>();
+    if (!propValue->expression->is<IR::ConstructorCallExpression>()) {
+        ::error("Table '%1%' has an implementation which is not a constructor call: %2%",
+                table->externalName(), impl);
+        return boost::none;
+    }
+    const auto constructorCallExpr =
```

Comment:
I'd be happy to make that change, but could you provide an example of a P4 program where that's true? Every P4 program I've tried, both P4-14 and P4-16, has only had a constructor call here by the time this code runs. In P4-14, a reference to a Declaration_Instance seems like it would make more sense, but we seem to be rewriting them into constructor calls.